### DISCUSSION TITLE
Confusing doc: building docker image with custom provider(s)

### DISCUSSION THREAD

**Discussion by erickhagstrom** [keycloak/keycloak#22004] [Q&A] [answered]
I want to build a Docker image that contains a custom provider, so of course I read the documentation. In the Server Guide for Running Keycloak in a container (https://www.keycloak.org/server/containers) (https://github.com/keycloak/keycloak/blob/main/docs/guides/server/containers.adoc) I find the following line:

> To install custom providers, you just need to define a step to include the JAR file(s) into the /opt/keycloak/providers directory:
>
> \# A example build step that downloads a JAR file from a URL and adds it to the providers directory
>ADD --chown=keycloak:keycloak <MY_PROVIDER_JAR_URL> /opt/keycloak/providers/myprovider.jar

In other words, that line should be in my Dockerfile _somewhere_. But where? Specifically, does it go in the first section of the Dockerfile prior to running the `build` command? Or in the second section, after I've copied the newly built and optimized `/opt/keycloak/` into a fresh Keycloak image?

I'll gladly submit a pull request to make the documentation clear, but I don't know which is correct.

**Comment by vmuzikar** [comment, marked as answer]
The extensions need to be present before running the `build` command. I.e. you need to put it before this line:
https://github.com/keycloak/keycloak/blob/8ad6bc49a3903cfefcc26c251890c35de8fa050e/docs/guides/server/containers.adoc?plain=1#L37
A contribution is always most welcome. :)
